posts/ebpf-and-kernel-modules-whats-the-difference.md

```diff
@@ -0,0 +1,115 @@
+
+When kernel modules are loaded into the kernel, the kernel allocates memory for that program to execute, gives it all the privileges kernel code would get, and keeps a list tracking all kernel module names and locations so that the kernel can initialize and cleanup modules on user requests. Kernel module code runs directly on the CPU, like the kernel does.
+
+eBPF programs on the other hand will run as bytecode inside the eBPF VM, which is a sandboxed subsystem of the linux kernel. The bytecode is Just-in-Time (JIT) compiled into machine code during execution. This is similar to how Java bytecode runs on the JVM. The eBPF Linux subsystem will also keep track of running programs. 
+
+
+# Installing Applications
+
+When installing existing eBPF or Kernel Module programs, there's a good chance it will be installed just as any other Linux program is: via a package manager, a distributed binary, or an install script.
+
+Cilium ships a client CLI binary that is used to install the Cilium eBPF components onto kubernetes clusters. Falco is shipped as a package for both the eBPF and kernel module drivers. Wireguard is shipped as a package. Some kernel modules are shipped in Linux distributions and then can be enabled on install.
+
+Installing from source is a bit different than other programs.
+
+In eBPF's case, eBPF development packages are needed for compilation. Once the eBPF program is compiled, it will most likely be loaded into the eBPF kernel subsystem with the help of a user-space application that loads and manages the program. The user-space code will also likely be distributed with the eBPF code.
+
+In Kernel Module's case, kernel module development packages are needed for compilation. The build system is typically written in a Makefile. Once the binary is compiled, it will need to be loaded with kernel module management command `insmod`.
+
+
+# Supported Programming Languages
+eBPF programs can be written in any language that has a compiler that can output eBPF bytecode. Currently, and as far as I know, those languages are C, Rust, [BPFTrace](https://github.com/iovisor/bpftrace) and [P4](https://github.com/p4lang/p4c/tree/main).
+
+Kernel modules can be written in C or Rust.
+
+As mentioned, eBPF programs are usually launched with user-space applications. These applications can be written in C, Rust, Go, Python, or any language that has libraries that can load eBPF binaries into the eBPF linux subsystem.
+
+# Development Rigidity
+Kernel Modules are very general purpose and can be used to write any type of program. There is no hand-holding or guardrails either. 
+
+eBPF on the other hand is very rigid and structured. There are even different specified eBPF program types. Each program type will then be restricted on the types of data or functions they are allowed to access since each program is geared towards a narrow domain. To get an eBPF program to compile, it must also pass a verifier check for program correctness and complexity.
```

Comment:
I think it would be cool to add more about this "program correctness and complexity" checker, could probably be a whole separate blog post.  Really cool stuff!!!